nano/node/election.cpp

```diff
@@ -528,6 +535,83 @@ void nano::election::remove_votes (nano::block_hash const & hash_a)
 	}
 }
 
+void nano::election::remove_block (nano::block_hash const & hash_a)
+{
+	debug_assert (!mutex.try_lock ());
+	debug_assert (status.winner->hash () != hash_a);
+	if (auto existing = last_blocks.find (hash_a); existing != last_blocks.end ())
+	{
+		for (auto i (last_votes.begin ()); i != last_votes.end ();)
+		{
+			if (i->second.hash == hash_a)
+			{
+				i = last_votes.erase (i);
+			}
+			else
+			{
+				++i;
+			}
+		}
+		node.network.publish_filter.clear (existing->second);
+		last_blocks.erase (hash_a);
+	}
+}
+
+bool nano::election::replace_by_weight (nano::unique_lock<std::mutex> & lock_a, nano::block_hash const & hash_a)
+{
+	debug_assert (lock_a.owns_lock ());
+	nano::block_hash replaced_block (0);
+	auto winner_hash (status.winner->hash ());
+	// Sort existing blocks tally
+	std::vector<std::pair<nano::block_hash, nano::uint128_t>> sorted;
+	sorted.reserve (last_tally.size ());
+	std::copy (last_tally.begin (), last_tally.end (), std::back_inserter (sorted));
+	// Sort in ascending order
+	std::sort (sorted.begin (), sorted.end (), [](auto const & left, auto const & right) { return left.second < right.second; });
+	lock_a.unlock ();
```

Comment:
Seems like it could be unlocked before sorting

Reply:
Updated